tests/kernel/nested_irq/src/main.c

```diff
@@ -0,0 +1,93 @@
+#include <ztest.h>
+#define DURATION 5
+struct k_timer timer;
+
+#define ISR0_OFFSET 1
+#define ISR1_OFFSET 2
+
+#define IRQ_LINE(offset) (CONFIG_NUM_IRQS - ((offset) + 1))
+#define NUM_SECONDS(x) ((x) * 1000)
+
+volatile u32_t val;
+u32_t VAL = 0xDEAD;
+
+void isr1(void *param)
+{
+	ARG_UNUSED(param);
+	val = 0xDEAD;
+	printk("%s ran !!\n", __func__);
+}
+
+#if defined(CONFIG_ARM)
+#include <arch/arm/cortex_m/cmsis.h>
+void trigger_irq(int irq)
+{
+	printk("Triggering irq : %d\n", irq);
+#if defined(CONFIG_SOC_TI_LM3S6965_QEMU) || defined(CONFIG_CPU_CORTEX_M0) \
+	|| defined(CONFIG_CPU_CORTEX_M0PLUS)
+	NVIC_SetPendingIRQ(irq);
+#else
+	NVIC->STIR = irq;
+#endif
+}
+
+#elif defined(CONFIG_RISCV32) && !defined(CONFIG_SOC_RISCV32_PULPINO)
+void trigger_irq(int irq)
+{
+	u32_t mip;
+
+	printk("Triggering irq : %d\n", irq);
+	__asm__ volatile ("csrrs %0, mip, %1\n"
+			: "=r" (mip)
+			: "r" (1 << irq));
+}
+
+#elif defined(CONFIG_CPU_ARCV2)
+void trigger_irq(int irq)
+{
+	printk("Triggering irq : %d\n", irq);
+	_arc_v2_aux_reg_write(_ARC_V2_AUX_IRQ_HINT, irq);
+}
+#else
+/* So far, Nios II does not support this */
+void trigger_irq(int irq)
+{
+	ARG_UNUSED(irq);
+}
+#endif
+
+
+static void handler(struct k_timer *timer)
+{
+	ARG_UNUSED(timer);
+	irq_enable(IRQ_LINE(ISR1_OFFSET));
+	trigger_irq(IRQ_LINE(ISR1_OFFSET));
+}
+
+void isr0(void *param)
+{
+	ARG_UNUSED(param);
+	printk("%s running !!\n", __func__);
+	k_busy_wait(NUM_SECONDS(10));
+	printk("%s execution completed !!\n", __func__);
+	zassert_equal(val, VAL, "Nested interrupt is not working\n");
+}
+
+void test_nested_isr(void)
+{
+	IRQ_CONNECT(IRQ_LINE(ISR0_OFFSET), 1, isr0, NULL, 0);
+	IRQ_CONNECT(IRQ_LINE(ISR1_OFFSET), 0, isr1, NULL, 0);
+
```

Comment:
Some comments on what is being done here would be helpful. It appears you are using a timer just to trigger isr1 while isr0 is in a busy_wait. If indeed that is the case, you should mention this in comments.

Reply:
Addressed your comment and pushed new patch. Please have a look it.